samples/demo-server/src/customs.rs

```diff
@@ -0,0 +1,199 @@
+        description: name.into(),
+        display_name: name.into(),
+        value,
+    }
+}
+
+fn add_enum_data_type(nm: &Arc<SimpleNodeManager>, ns: u16) -> NodeId {
+    let mut addr = nm.address_space().write();
+
+    let id = NodeId::next_numeric(ns);
+
+    let type_def = DataTypeDefinition::Enum(EnumDefinition {
+        fields: Some(vec![
+            enum_field("Disabled", 1),
+            enum_field("Enabled", 2),
+            enum_field("Idle", 3),
+            enum_field("MoveAbs", 4),
+            enum_field("Error", 5),
+        ]),
+    });
+    DataTypeBuilder::new(&id, "AxisState", "AxisState")
+        .organized_by(DataTypeId::Enumeration)
+        .data_type_definition(type_def)
+        .insert(&mut *addr);
+
+    id
+}
+
+fn add_encoding(nm: &SimpleNodeManager, ns: u16, struct_id: &NodeId) -> NodeId {
+    let mut addr = nm.address_space().write();
+    let id = NodeId::next_numeric(ns);
+    ObjectBuilder::new(&id, "DefaultBinary", "DefaultBindary")
+        .reference(
+            struct_id,
+            ReferenceTypeId::HasEncoding,
+            ReferenceDirection::Inverse,
+        )
+        .insert(&mut *addr);
+    id
+}
+
+fn add_custom_data_type(nm: &SimpleNodeManager, ns: u16, e_state_id: &NodeId) -> NodeId {
+    let struct_id = NodeId::next_numeric(ns);
+    let enc_id = add_encoding(nm, ns, &struct_id);
+
+    let type_def = DataTypeDefinition::Structure(StructureDefinition {
+        default_encoding_id: NodeId::null(),
+        base_data_type: DataTypeId::Structure.into(),
+        structure_type: StructureType::Structure,
+        fields: Some(vec![
+            StructureField {
+                name: "sErrorMessage".into(),
+                data_type: DataTypeId::String.into(),
+                value_rank: -1,
+                ..Default::default()
+            },
+            StructureField {
+                name: "nErrorID".into(),
+                data_type: DataTypeId::UInt32.into(),
+                value_rank: -1,
+                ..Default::default()
+            },
+            StructureField {
+                name: "eLastState".into(),
+                data_type: e_state_id.clone(),
+                value_rank: -1,
+                ..Default::default()
+            },
+        ]),
+    });
+    let mut addr = nm.address_space().write();
+    DataTypeBuilder::new(&struct_id, "ErrorData", "ErrorData")
+        .organized_by(DataTypeId::Structure)
+        .data_type_definition(type_def)
+        .reference(
+            enc_id,
+            ReferenceTypeId::HasEncoding,
+            ReferenceDirection::Forward,
+        )
+        .insert(&mut *addr);
+
+    struct_id
+}
+
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+#[repr(i32)]
+pub enum AxisState {
+    Disabled = 1i32,
+    Enabled = 2i32,
+    Idle = 3i32,
+    MoveAbs = 4i32,
+    Error = 5i32,
+}
+
+impl Default for AxisState {
+    fn default() -> Self {
+        Self::Disabled
+    }
+}
+
+impl TryFrom<i32> for AxisState {
```

Comment:
the entire i32 convertoin thing is really dum code. Can I use something like int-enum crate to do that?

Reply:
Making `derive(BinaryEncodable, BinaryDecodable)```support `repr` enums is on my TODO. That would be the cleanest way to do that IMO.

Reply:
Maybe a special trait for numeric enums with blanket implementations for ``IntoVariant`, `TryFrom`, `BinaryEncodable`, etc...

Reply:
that is exactly what the crate int-enum does. It has a trait IntEnum which does exactly that. Not tested for that particular case yet but I will.

Reply:
Yeah, but we have slightly special requirements, and we probably want some other stuff as well to comply with how the standard demands we name enum variants.

I'll look into it, and look at `int-enum` as well.

Reply:
It also strikes me that blanket impls won't work, we still need to derive `BinaryEncodable`, etc. in this case. There's a bigger effort to be done there to _really_ extend the derive macros to cover more of the standard like unions, structs with optional types, and these enums...